File: proxyshield/src/algorithms/entropy.js
/**
 * Calculate Shannon entropy of a string to detect encoded or obfuscated payloads.
 * Normal English text has entropy around 3.5–4.5 bits per byte.
 * Base64-encoded attack payloads have entropy around 5.5–6.0.
 *
 * @param {string} text - The input string to analyze
 * @returns {number} Entropy value in bits per byte (range 0 to 8)
 */
export function calculateEntropy(text) {
  if (!text || text.length === 0) return 0;

  const freq = new Map();
  for (const char of text) {
    freq.set(char, (freq.get(char) || 0) + 1);
  }

  const total = text.length;
  let entropy = 0;

  for (const count of freq.values()) {
    const p = count / total;
    if (p > 0) {
      entropy -= p * Math.log2(p);
    }
  }

  return entropy;
}

File: proxyshield/proxyshield-core/internal/algorithm/entropy.go
// Package algorithm provides rate limiting algorithms and entropy calculation.
package algorithm

import "math"

// CalculateEntropy computes the Shannon entropy of the given text in bits per byte.
// Returns a value between 0 and 8. Normal English text scores 3.5-4.5.
// Base64-encoded payloads score 5.5-6.0.
// Uses a fixed [256]int frequency array — no heap allocation.
func CalculateEntropy(data []byte) float64 {
	if len(data) == 0 {
		return 0
	}

	var freq [256]int
	for _, b := range data {
		freq[b]++
	}

	total := float64(len(data))
	entropy := 0.0

	for _, count := range freq {
		if count == 0 {
			continue
		}
		p := float64(count) / total
		entropy -= p * math.Log2(p)
	}

	return entropy
}
